[Flang][OpenMP] Add Parsing support for Indirect Clause

Stylie777 opened the pull request:
As part of OpenMP Version 5.1, support for the `indirect` clause was added for the `declare target` directive. This clause should follow an `enter` clause, and allows procedure calls to be done indirectly through OpenMP.

This adds Parsing support for the clause, along with semantics checks. Currently, lowering for the clause is not supported so a TODO message will be outputted to the user. It also performs version checking as `indirect` is only support in OpenMP 5.1 or greater.

See also: #110008

Stylie777 committed:
[Flang][OpenMP] Add Parsing support for Indirect Clause

As part of OpenMP Version 5.1, support for the `indirect` clause
was added for the `declare target` directive. This clause should
follow an `enter` clause, and allows procedure calls to be done
indirectly through OpenMP.

This adds Parsing support for the clause, along with semantics
checks. Currently, lowering for the clause is not supported so
a TODO message will be outputted to the user. It also performs
version checking as `indirect` is only support in OpenMP 5.1 or
greater.

See also: #110008

Stylie777 committed:
Apply formatting changes

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -1812,15 +1812,25 @@ void OmpStructureChecker::Leave(const parser::OmpDeclareTargetWithClause &x) {
     }
+    if (indirectClause && version < 51) {
+      context_.Say(x.source,
+          "The usage of INDIRECT clause on DECLARE TARGET directive is only supported on OpenMP Version 5.1 or greater"_err_en_US);
```
We use a slightly different form of "feature not supported in version xyz" messages, see [lines 278-281](https://github.com/llvm/llvm-project/blob/main/flang/lib/Semantics/check-omp-structure.cpp#L278-L281) for example.  The helper functions ThisVersion and TryVersion generate parts of the message.

tblah replied:
Maybe CheckAllowedClause could be used for this?

Stylie777 committed:
Responding to review comments

The following has been completed:
- Removed unneeded function declartion
- Updated logic for version checking to use the `CheckAllowedClause`
function.

tblah reviewed (approved):
LGTM, thanks. But please wait for approval from @kparzysz
